File: LeetCode/238. Product of Array Except Self/elixir/solution.exs
defmodule Solution do
  @spec product_except_self(nums :: [integer]) :: [integer]
  def product_except_self(nums) do
    {product, zero_count} =
      Enum.reduce(nums, {1, 0}, fn x, {acc, zero_count} ->
        if x == 0 do
          {acc, zero_count + 1}
        else
          {x * acc, zero_count}
        end
      end)

    Enum.map(nums, &div(product, &1, zero_count))
  end

  defp div(a, b, 0), do: trunc(a / b)
  defp div(a, b, 1), do: if(b == 0, do: a, else: 0)
  defp div(_a, _b, _zero_count), do: 0
end

File: LeetCode/238. Product of Array Except Self/java/Solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int product = 1;
        int productWithOutZero = 1;
        int zeroCount = 0;
        for (int num : nums) {
            product *= num;
            if (num != 0) {
                productWithOutZero *= num;
            } else {
                zeroCount++;
            }
        }
        int length = nums.length;
        int[] answer = new int[length];
        for (int i = 0; i < length; i++) {
            int num = nums[i];
            if (zeroCount == 0) {
                answer[i] = product / num;
            } else {
                if (num == 0 && zeroCount == 1) {
                    answer[i] = productWithOutZero;
                } else {
                    answer[i] = 0;
                }
            }
        }
        return answer;
    }
}